File: apps/desktop/src/lib/greeks.ts
/**
 * Black-Scholes-Merton (dividend-free) option pricing with the full first- and
 * second-order Greek set. Ported verbatim from the general.exchange web project
 * `greeksService.ts` so the desktop terminal reports identical numbers.
 *
 * Conventions:
 *  - vega is per 1% (0.01) change in volatility
 *  - theta, charm, and color are reported per calendar day
 *  - all other Greeks are per unit / per year as standard
 */

export type ContractType = 'CALL' | 'PUT';

export interface OptionContract {
  strike: number;
  underlyingPrice: number;
  riskFreeRate: number;
  impliedVolatility: number;
  /** time to expiration in years */
  timeToExpiration: number;
  type: ContractType;
}

export interface FullGreeks {
  delta: number;
  gamma: number;
  theta: number; // per day
  vega: number; // per 1% vol
  rho: number; // per 1% rate
  lambda: number; // elasticity
  epsilon: number; // dividend rho (q = 0)
  charm: number; // dDelta/dTime, per day
  vanna: number; // dDelta/dVol
  volga: number; // dVega/dVol (vomma)
  speed: number; // dGamma/dSpot
  zomma: number; // dGamma/dVol
  color: number; // dGamma/dTime, per day
  price: number; // theoretical option price
}

const YEAR = 365;

/** Abramowitz & Stegun error-function approximation. */
function erf(x: number): number {
  const a1 = 0.254829592;
  const a2 = -0.284496736;
  const a3 = 1.421413741;
  const a4 = -1.453152027;
  const a5 = 1.061405429;
  const p = 0.3275911;

  const sign = x >= 0 ? 1 : -1;
  const ax = Math.abs(x);
  const t = 1.0 / (1.0 + p * ax);
  const y = 1.0 - ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-ax * ax);
  return sign * y;
}

const N = (x: number) => 0.5 * (1 + erf(x / Math.SQRT2));
const phi = (x: number) => Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);

export function calculateFullGreeks(option: OptionContract): FullGreeks {
  const { underlyingPrice: S, strike: K, timeToExpiration: T, riskFreeRate: r, impliedVolatility: sigma, type } = option;
  const isCall = type === 'CALL';

  if (T <= 0 || sigma <= 0) {
    const delta = isCall ? (S > K ? 1 : 0) : S < K ? -1 : 0;
    return {
      delta, gamma: 0, theta: 0, vega: 0, rho: 0, lambda: 0, epsilon: 0,
      charm: 0, vanna: 0, volga: 0, speed: 0, zomma: 0, color: 0,
      price: Math.max(isCall ? S - K : K - S, 0),
    };
  }

  const sqrtT = Math.sqrt(T);
  const d1 = (Math.log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * sqrtT);
  const d2 = d1 - sigma * sqrtT;
  const disc = Math.exp(-r * T);

  const delta = isCall ? N(d1) : N(d1) - 1;
  const gamma = phi(d1) / (S * sigma * sqrtT);
  const vega = (S * phi(d1) * sqrtT) / 100;
  const thetaYr = isCall
    ? -(S * phi(d1) * sigma) / (2 * sqrtT) - r * K * disc * N(d2)
    : -(S * phi(d1) * sigma) / (2 * sqrtT) + r * K * disc * N(-d2);
  const theta = thetaYr / YEAR;
  const rho = (isCall ? K * T * disc * N(d2) : -K * T * disc * N(-d2)) / 100;

  const price = isCall ? S * N(d1) - K * disc * N(d2) : K * disc * N(-d2) - S * N(-d1);
  const lambda = price !== 0 ? delta * (S / price) : 0;
  const epsilon = isCall ? -S * T * N(d1) : S * T * N(-d1);

  const charmYr = -phi(d1) * ((2 * r * T - d2 * sigma * sqrtT) / (2 * T * sigma * sqrtT));
  const charm = charmYr / YEAR;
  const vanna = -phi(d1) * (d2 / sigma);
  const volga = vega * ((d1 * d2) / sigma);
  const speed = -(gamma / S) * (d1 / (sigma * sqrtT) + 1);
  const zomma = gamma * ((d1 * d2 - 1) / sigma);
  const colorYr =
    (-phi(d1) / (2 * S * T * sigma * sqrtT)) *
    (2 * r * T + 1 + (d1 * (2 * r * T - d2 * sigma * sqrtT)) / (sigma * sqrtT));
  const color = colorYr / YEAR;

  return { delta, gamma, theta, vega, rho, lambda, epsilon, charm, vanna, volga, speed, zomma, color, price };
}

/** Convenience: years remaining until an ISO expiration date. */
export function yearsToExpiration(expirationIso: string, now = Date.now()): number {
  const exp = new Date(expirationIso).getTime();
  return Math.max(0, (exp - now) / (YEAR * 24 * 60 * 60 * 1000));
}

File: backend/rust/crates/gx-core/src/greeks.rs
//! Direct port of `backend/python/common/greeks.py`.
//! Conventions: vega per 1% vol; theta/charm/color per calendar day.

use statrs::distribution::{Continuous, ContinuousCDF, Normal};

const YEAR: f64 = 365.0;

fn norm() -> Normal {
    Normal::new(0.0, 1.0).unwrap()
}

fn ncdf(x: f64) -> f64 {
    norm().cdf(x)
}

fn npdf(x: f64) -> f64 {
    norm().pdf(x)
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Greeks {
    pub price: f64,
    pub delta: f64,
    pub gamma: f64,
    pub theta: f64,
    pub vega: f64,
    pub rho: f64,
    #[serde(rename = "lambda")]
    pub lambda_: f64,
    pub epsilon: f64,
    pub charm: f64,
    pub vanna: f64,
    pub volga: f64,
    pub speed: f64,
    pub zomma: f64,
    pub color: f64,
}

/// Compute BSM Greeks. `opt_type`: `"CALL"` or `"PUT"` (case-insensitive).
pub fn bsm(
    spot: f64,
    strike: f64,
    t_years: f64,
    rate: f64,
    sigma: f64,
    opt_type: &str,
) -> Greeks {
    let is_call = opt_type.to_uppercase() == "CALL";

    if t_years <= 0.0 || sigma <= 0.0 {
        let intrinsic = if is_call {
            (spot - strike).max(0.0)
        } else {
            (strike - spot).max(0.0)
        };
        let delta = if is_call && spot > strike {
            1.0
        } else {
            0.0
        };
        return Greeks {
            price: intrinsic,
            delta,
            gamma: 0.0,
            theta: 0.0,
            vega: 0.0,
            rho: 0.0,
            lambda_: 0.0,
            epsilon: 0.0,
            charm: 0.0,
            vanna: 0.0,
            volga: 0.0,
            speed: 0.0,
            zomma: 0.0,
            color: 0.0,
        };
    }

    let sqrt_t = t_years.sqrt();
    let d1 =
        ((spot / strike).ln() + (rate + 0.5 * sigma * sigma) * t_years) / (sigma * sqrt_t);
    let d2 = d1 - sigma * sqrt_t;
    let disc = (-rate * t_years).exp();
    let phi = npdf(d1);

    let (price, delta, theta_yr, rho, epsilon) = if is_call {
        let nd1 = ncdf(d1);
        let nd2 = ncdf(d2);
        let p = spot * nd1 - strike * disc * nd2;
        let th = -(spot * phi * sigma) / (2.0 * sqrt_t) - rate * strike * disc * nd2;
        let r = strike * t_years * disc * nd2 / 100.0;
        let e = -spot * t_years * nd1;
        (p, nd1, th, r, e)
    } else {
        let delta = ncdf(d1) - 1.0;
        let nnd1 = ncdf(-d1);
        let nnd2 = ncdf(-d2);
        let p = strike * disc * nnd2 - spot * nnd1;
        let th = -(spot * phi * sigma) / (2.0 * sqrt_t) + rate * strike * disc * nnd2;
        let r = -strike * t_years * disc * nnd2 / 100.0;
        let e = spot * t_years * nnd1;
        (p, delta, th, r, e)
    };

    let gamma = phi / (spot * sigma * sqrt_t);
    let vega = spot * phi * sqrt_t / 100.0;
    let lambda_ = if price.abs() > 1e-12 {
        delta * (spot / price)
    } else {
        0.0
    };
    let charm_yr = -phi * ((2.0 * rate * t_years - d2 * sigma * sqrt_t)
        / (2.0 * t_years * sigma * sqrt_t));
    let color_yr = -(phi / (2.0 * spot * t_years * sigma * sqrt_t))
        * (2.0 * rate * t_years
            + 1.0
            + (d1 * (2.0 * rate * t_years - d2 * sigma * sqrt_t)) / (sigma * sqrt_t));

    Greeks {
        price,
        delta,
        gamma,
        theta: theta_yr / YEAR,
        vega,
        rho,
        lambda_,
        epsilon,
        charm: charm_yr / YEAR,
        vanna: -phi * (d2 / sigma),
        volga: vega * (d1 * d2 / sigma),
        speed: -(gamma / spot) * (d1 / (sigma * sqrt_t) + 1.0),
        zomma: gamma * ((d1 * d2 - 1.0) / sigma),
        color: color_yr / YEAR,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn atm_call_sanity() {
        let g = bsm(100.0, 100.0, 0.25, 0.05, 0.2, "CALL");
        assert!(
            (g.price - 5.1).abs() < 0.5,
            "ATM call price out of range: {}",
            g.price
        );
        assert!((g.delta - 0.5).abs() < 0.1, "ATM delta: {}", g.delta);
        assert!(g.gamma > 0.0 && g.vega > 0.0);
    }

    #[test]
    fn put_call_parity() {
        let c = bsm(100.0, 100.0, 1.0, 0.05, 0.3, "CALL");
        let p = bsm(100.0, 100.0, 1.0, 0.05, 0.3, "PUT");
        let lhs = c.price - p.price;
        let rhs = 100.0 - 100.0 * (-0.05_f64).exp();
        assert!(
            (lhs - rhs).abs() < 1e-4,
            "put-call parity violated: {lhs} vs {rhs}"
        );
    }
}
